File: qpid/cpp/bindings/qpid/examples/perl/spout.pl
#!/usr/bin/env perl
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
use strict;
use warnings;

use qpid;
use Getopt::Long;
use Pod::Usage;
use Time::Local;

my $url     = "127.0.0.1";
my $timeout = 0;
my $count   = 1;
my $durable = 0;
my $id      = "";
my $replyto = "";
my @properties;
my @entries;
my $content           = "";
my $connectionOptions = "";
my $address           = "amq.direct";
my $help;

my $result = GetOptions(
    "broker|b=s"           => \$url,
    "timeout|t=i"          => \$timeout,
    "count|c=i"            => \$count,
    "durable|d"            => \$durable,
    "id|i=s"               => \$id,
    "replyto=s"            => \$replyto,
    "property|p=s@"        => \@properties,
    "map|m=s@"             => \@entries,
    "content=s"            => \$content,
    "connection-options=s" => \$connectionOptions,
    "help|h"               => \$help
) || pod2usage( -verbose => 0 );

pod2usage( -verbose => 1 ) if $help;

if ( $#ARGV ge 0 ) {
    $address = $ARGV[0];
}

sub setEntries {
    my ($content) = @_;

    foreach (@entries) {
        my ( $name, $value ) = split( "=", $_ );
        $content->{$name} = $value;
    }
}

sub setProperties {
    my ($message) = @_;

    foreach (@properties) {
        my ( $name, $value ) = split( "=", $_ );
        $message->set_property( $name, $value );
    }
}

# create a connection object
my $connection = new qpid::messaging::Connection( $url, $connectionOptions );

eval {
    # open the connection, create a session and then a sender
    $connection->open();
    my $session = $connection->create_session();
    my $sender  = $session->create_sender($address);

    # create a message to be sent
    my $message = new qpid::messaging::Message();
    setProperties($message) if (@properties);
    if (@entries) {
        my $content = {};
        setEntries($content);
        $message->set_content_object($content);
    }
    elsif ($content) {
        $message->set_content($content);
        $message->set_content_type("text/plain");
    }

    # set durable flag
    $message->set_durable($durable);

    # if a reply-to address was supplied, then create a receiver from the
    # session and wait for a response to be sent
    my $receiver;
    if ($replyto) {
        my $responseQueue = new qpid::messaging::Address($replyto);
        $receiver = $session->create_receiver($responseQueue);
        $message->set_reply_to($responseQueue);
    }

    my $start = localtime;
    my @s     = split( /[:\s]/, $start );
    my $s     = "$s[3]$s[4]$s[5]";
    my $n     = $s;

    for (
        my $i = 0 ;
        ( $i < $count || $count == 0 )
          and ( $timeout == 0 || abs( $n - $s ) < $timeout ) ;
        $i++
      )
    {

        $sender->send($message);

        if ($receiver) {
            print "Waiting for a response.\n";
            my $response = $receiver->fetch();
            print "$i -> " . $response->get_content() . "\n";
        }

        my $now = localtime;
        my @n   = split( /[:\s]/, $now );
        my $n   = "$n[3]$n[4]$n[5]";
    }
    $session->sync();
    $connection->close();
};

if ($@) {
    $connection->close();
    die $@;
}

__END__

=head1 NAME

spout - Send messages to the specified address

=head1 SYNOPSIS

  Usage: spout [OPTIONS] ADDRESS

  Options:
  -h, --help                   show this message
  -b VALUE, --broker VALUE     url of broker to connect to
  -t VALUE, --timeout VALUE    exit after the specified time
  -c VALUE, --count VALUE      stop after count messageshave been sent, zero disables
  -i VALUE, --id VALUE         use the supplied id instead of generating one
  --replyto VALUE              specify reply-to value
  -P VALUE, --property VALUE   specify message property
  -M VALUE, --map VALUE        specify entry for map content
  --content VALUE              specify textual content
  --connection-options VALUE   connection options string in the form {name1:value1, name2:value2}

=cut

File: qpid/cpp/bindings/qpid/ruby/examples/spout.rb
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

$:.unshift File.join(File.dirname(__FILE__), "..", "lib")

require 'qpid_messaging'
require 'optparse'

options = {
  :broker => "127.0.0.1",
  :address => "",
  :timeout => 0,
  :durable => false,
  :count => 1,
  :properties => {},
  :content => nil,
  :mapped => {}
}

opts = OptionParser.new do |opts|
  opts.banner = "Usage: spout.rb [OPTIONS] ADDRESS"

  opts.on("-h", "--help",
          "show this message") do |help|
    puts opts
    exit
  end

  opts.on("-b","--broker VALUE",
          "url of broker to connect to ") do |broker|
    options[:broker] = broker
  end

  opts.on("-t", "--timeout VALUE", Integer,
          "exit after the specified time") do |timeout|
    options[:timeout] = Qpid::Messaging::Duration.new timeout * 1000
  end

  opts.on("-d", "--durable",
          "make the message durable (def. #{options[:durable]})") do
    options[:durable] = true
  end

  opts.on("-c", "--count VALUE", Integer,
          "stop after count messages have been sent, zero disables") do |count|
    options[:count] = count
  end

  opts.on("-i", "--id VALUE",
          "use the supplied id instead of generating one") do |id|
    options[:id] = id
  end

  opts.on("--reply-to VALUE",
          "specify reply-to address") do |replyto|
    options[:replyto] = replyto
  end

  opts.on("-P", "--property VALUE",
          "specify message property") do |property|
    name  = property.split(/=/)[0]
    value = property.split(/=/)[1]
    options[:properties][name] = value
  end

  opts.on("-M", "--map VALUE",
          "specify entry for map content") do |mapped|
    name  = mapped.split(/=/)[0]
    value = mapped.split(/=/)[1]
    options[:mapped][name] = value
  end

  opts.on("--content VALUE",
          "specify textual content") do |content|
    options[:content] = content
  end

  opts.on("--connection-options VALUE",
          "connection options string in the form {name1:value1, name2:value2}") do |conopts|

    options[:connection_options] = conopts
  end
end

begin
  opts.parse!(ARGV)
rescue => error
  opts.parse(["-h"])
end

# now get the non-arg options
options[:address] = ARGV[0] unless ARGV[0].nil?

# process the connection options
unless options[:connection_options].nil?
  fields = options[:connection_options].gsub(/^\{(.*)\}$/, '\1')
  # remove any surrounding braces
  if /\{.*\}/ =~ fields
    fields = fields[1..-2]
  end
  # break up the options separated by commas
  keysvalues = {}
  fields.split(",").each do |field|
    if /.+:.+/ =~ field
      (key, value) = field.split(":")
      keysvalues[key] = value
    end
  end
  # now store the options
  options[:connection_options] = keysvalues
end

connection = Qpid::Messaging::Connection.new(:url => options[:broker],
                                             :options => options[:connection_options])
connection.open
session = connection.create_session
sender = session.create_sender options[:address]
message = Qpid::Messaging::Message.new

options[:properties].each_key {|key| message[key] = options[:properties][key]}

(1..options[:count]).each do |count|
  if !options[:mapped].keys.empty?
    message.content_object = options[:mapped]
  elsif options[:content]
    message.content_object = options[:content]
  end
  message.durable = options[:durable]
  message.content_object = options[:content] unless options[:content].nil?
  message.properties["spout-id"] = "#{count}"
  message.reply_to = options[:replyto] unless options[:replyto].nil? || options[:replyto].empty?
  sender.send message
end

# session.sync

connection.close

